### ISSUE TITLE
Scale/Translate TileLayer

### ISSUE THREAD

**Issue by ilan-gold** [visgl/deck.gl#5047]
#### Target Use case
Much like you can scale/translate a `BitmapLayer` by affecting its bounds, it would be cool to do the same to the `TileLayer`.  The hairy part would be the interplay between the `Viewport` and tile fetching but I think it could be possible.  Is this something we'd be interested in supporting?  @kylebarron I would be happy to work on this PR (and make sure it's even possible).

#### Proposed feature
Add `scale` and `translate` parameters to the `TileLayer` to allow for uniform transformations to the layer and and underlying tiles.

#### To Do List 
  - [ ] Add label and assign to milestone 
  - [ ] Coding
  - [ ] Doc update
  - [ ] What’s new update 
  - [ ] Test

**Comment by Pessimistress** [comment]
Does `modelMatrix` not work?

**Comment by ilan-gold** [comment]
@Pessimistress are you saying this would be how it is implemented internally to expose an API or you could use this to get the same effect without the change?

**Comment by Pessimistress** [comment]
[modelMatrix](https://deck.gl/docs/api-reference/core/layer#modelmatrix) should be supported by all layers. If it doesn't work then it's a bug. Unless `TileLayer` can justify a different use case, there is no need for a new API.

**Comment by ilan-gold** [comment]
@Pessimistress But don't we need this to be factored into viewport we use while fetching tiles as well?  I added this to the `props` of the `image-tile` example but then it only seems to affect rendering and not fetching

![ezgif com-video-to-gif(9)](https://user-images.githubusercontent.com/43999641/96642052-9587f400-12f3-11eb-9961-bd61f3b1940e.gif)

The docs you shared suggest that it should affect the `project` function in Javascript (not just on the shaders) but when I log out the `viewport` in the `getTileIndices` function, `modelMatrix` is `null`.  Is this then a bug, or should it propagate only when a new `viewport` is made?  Should we be creating a new `viewport` with the `modelMatrix` every time we get indices (instead of using `this.context.viewport`)?

**Comment by Pessimistress** [comment]
It is the expected behavior. `viewport` is global, while `modelMatrix` only affects one layer.

I don't think I understand your use case. Why do you need to shift the index system?

**Comment by ilan-gold** [comment]
@Pessimistress The idea would be to have multiple `TileLayer`s from different sources laid out on a grip so that you can zoom out and see all the little top-of-the-image-pyramid thumbnails, but then zoom in on individual ones, like holding a picture frame up to your eye.

**Comment by Pessimistress** [comment]
That sounds really hacky to me. I think a reasonable approach is to make a virtual indexing system for a single TileLayer (pseudocode):

```js
const TILE_SOURCE_A = {
  url: `http://a.server.com/{z}/{x}/{y}.png`,
  bounds: [-100, 0, 0, 100],
  translate: [-100, 0]
};
const TILE_SOURCE_B = {
  url: `http://b.server.com/{z}/{x}/{y}.png`,
  bounds: [0, 0, 100, 100],
  translate: [0, 0]
};

function getTileData({x, y, z, bbox}) {
  let source;
  if (intersect(bbox, TILE_SOURCE_A.bounds)) {
    source = TILE_SOURCE_A;
  } else if (intersect(bbox, TILE_SOURCE_B.bounds)) {
    source = TILE_SOURCE_B;
  }
  if (source) {
    x -= source.translate[0] / (bbox.right - bbox.left);
    y -= source.translate[1] / (bbox.bottom - bbox.top);
    return fetchJSON(evaluate(source.url, x, y, z));
  }
}
```

**Comment by ilan-gold** [comment]
Yes I thought about this as well.  I am fine with doing this.  I think my thought was something along the lines of what used to exist in the `TileLayer`:

https://github.com/visgl/deck.gl/blob/ef42cf4e8340958824b53a75dc6dc42f333ad922/modules/geo-layers/src/tile-layer/utils/viewport-util.js#L32-L43

While this ended up in a somewhat faulty implementation for the `CARTESIAN` use case if memory serves, I don't remember it being the creation of a new viewport being problematic, but rather the order in which we created (i.e after cutting off to `Math.floor`).  So I guess my proposal would be to return to this paradigm but allow the `modelMatrix` to be applied as well in order to account for the fact that the `TileLayer` needs special handling to work properly with regards to fetching tiles.

I'm fine if we don't want to do this for whatever reason (your solution certainly works), but I would be happy to try it out and make a PR.

**Comment by ilan-gold** [comment]
@Pessimistress This is the motivating example for what I referenced: 
https://mystifying-lalande-e12142.netlify.app/?source=https://minio-dev.openmicroscopy.org/idr/idr0002-heriche-condensation/plate1_1_013/422.zarr

Ideally as you zoom in on the images, the resolution of them would in theory increase.  So my thought would be having a large number of `TileLayer` instances with `modelMatrix` `props` for each, managed by a `CompositeLayer`.  If you don't want to accept this as a feature we can close the issue - it might be worth noting then that `modelMatrix` does not work with `TileLayer` (I would guess the same is true for the 3D version as well).

**Comment by Pessimistress** [comment]
`modelMatrix` only applies to the LNGLAT coordinate system in a limited way (only scaling/offsetting z makes sense) and that still works with the current TileLayer implementation.

I am not opposed to supporting `modelMatrix` when using `Tileset2D` with non-geospatial coordinates. Feel free to open a PR if you get it working. My concerns are the complexity of this feature:

- How are you going to construct the transformed viewport, which could be generated by a number of View classes?
- Is it going to work with arbitrary transforms, e.g. rotation?

And if you are using this to render small multiples, memory usage will be a big issue. Each TileLayer will have its own cache.

**Comment by ilan-gold** [comment]
Thanks for the direction/pointers @Pessimistress.  I'll try to make progress soon.

**Comment by ilan-gold** [comment]
@Pessimistress I think have something working but I want to clarify my understanding first so I write/use the correct API.

It seems like the `unproject` function on a `viewport` object is not affected by the `modelMatrix` parameter of the viewport object - is this right? It does say in the docs that `unproject` only goes from screen to world space, so it makes sense to me as the `modelMatrix` is for model space, but I could be wrong.  [The documentation for `Layer`](https://deck.gl/docs/api-reference/core/layer#modelmatrix) states otherwise, that the `modelMatrix` does have get used by `unproject` but as we discussed, this usage does not propagate to `this.context.viewport` which the `TileLayer` uses. 

For example:

```javascript
let clone = Object.assign(Object.create(Object.getPrototypeOf(viewport)), viewport, {
    modelMatrix
});
isEqual(clone.unproject([10, 10]), viewport.unproject([10, 10])));
/// true
```

seems to be the case when I try it out.  In this case, I ended up using the (inverse of the) `modelMatrix` to achieve my desired effect but I am curious if I have this wrong.

**Comment by Pessimistress** [comment]
Viewport instances are not mutable. All transform matrices are calculated during construction.

`layer.project` and `layer.unproject` are more complicated than the methods on viewport. They deal with the coordinate system of the layer.

**Comment by ilan-gold** [comment]
Ah, ok @Pessimistress, I misunderstood the docs.  I see now.

**Comment by ilan-gold** [comment]
Would it be possible to do an alpha release soon to try this out? Thanks!

**Comment by Pessimistress** [comment]
@ilan-gold Published 8.4.0-alpha.2

**Comment by ilan-gold** [comment]
Thanks so much!!!!
